Allow streaming layered containers with non-root Nix store

the-sun-will-rise-tomorrow opened the pull request:
## Description of changes

This opens the way towards building images where Nix can be used as an unprivileged user (in single-user mode).

The following creates a container where you can run e.g. `nix-shell -p hello --run hello` as a non-root user:

```nix
{ system ? builtins.currentSystem
, nixpkgs ? <nixpkgs>}:
let
  pkgs = import nixpkgs { inherit system; };
  pkgsLinux = import nixpkgs { system = "x86_64-linux"; };
in
pkgs.dockerTools.streamLayeredImage rec {
  name = "non-root-nix";
  tag = "latest";
  uid = 1000;
  gid = 1000;
  uname = "user";
  gname = "user";
  contents = with pkgsLinux; [
    coreutils
    nix
  ];
  enableFakechroot = true;
  fakeRootCommands = ''
    set -eu
    ${pkgsLinux.dockerTools.shadowSetup}
    ${pkgsLinux.shadow}/bin/groupadd --gid ${toString gid} ${gname}
    ${pkgsLinux.shadow}/bin/useradd --create-home --uid ${toString uid} --gid ${toString gid} ${uname}
    mkdir /nix/var /nix/var/nix
    chown ${toString uid}:${toString gid} /nix /nix/var /nix/var/nix
    mkdir -m1777 /tmp
  '';
  config = {
    User = "user";
    Cmd = ["${pkgsLinux.bashInteractive}/bin/bash"];
    Env = [
      "NIX_PATH=nixpkgs=${nixpkgs}"
      "GIT_SSL_CAINFO=${pkgsLinux.cacert}/etc/ssl/certs/ca-bundle.crt"
      "SSL_CERT_FILE=${pkgsLinux.cacert}/etc/ssl/certs/ca-bundle.crt"
    ];
  };
}
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

the-sun-will-rise-tomorrow committed:
streamLayeredImage: Allow customizing ownership

This opens the way towards building images where Nix can be used as an
unprivileged user (in single-user mode).

the-sun-will-rise-tomorrow committed:
streamLayeredImage: Change mode of /nix, /nix/store to 755

The change is insignificant when the owner is root.  However, when it
is not root, this change is needed to allow using Nix (as an
unprivileged user) inside the container.

WxNzEMof commented:
Dependencies are in, so I think this is ready for review.

roberth commented on nixos/tests/docker-tools.nix:
This looks kinda ok, but it seems that we still need a test and/or assertion that checks file ownership.

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -856,6 +856,11 @@ rec {
     , # Time of creation of the image. Passing "now" will make the
       # created date be the time of building.
       created ? "1970-01-01T00:00:01Z"
+    , # Credentials for file ownership.
```
The real documentation is over at `doc/build-helpers/images/dockertools.section.md`.
(I think we should remove these comments and refer to that file instead.)

Which files are affected by this?

Probably the answer also answers:
Does it interact with `fakeRootCommands`?

roberth reviewed (commented):
The docs and tests need a little work, but the overall idea seems good.

the-sun-will-rise-tomorrow committed:
Document the new streamLayeredImage parameters

the-sun-will-rise-tomorrow committed:
Remove the redundant comments from streamLayeredImage parameters

The proper place to describe them is the documentation, where they are
described thoroughly.

the-sun-will-rise-tomorrow committed:
dockerTools.tests.docker-tools: Add non-root streamLayeredImage test

WxNzEMof commented:
> This looks kinda ok, but it seems that we still need a test and/or assertion that checks file ownership.

Added a simple test.

> The real documentation is over at `doc/build-helpers/images/dockertools.section.md`.

Ah, updated.

> (I think we should remove these comments and refer to that file instead.)

Done.

> Which files are affected by this?

This change is only for the store. Everything else can be fixed up in `fakeRootCommands` cheaply.

(We can't just `chown -R` the `/nix/store` in `fakeRootCommands` because that unfortunately causes the final layer to contain a copy of the entire store, just with different permissions.)

> Does it interact with `fakeRootCommands`?

No, in fact you will probably need to use it to make use of this, as in the example in the issue description.

WxNzEMof commented:
By the way, I'm guessing it's due to something with my machine, but the `"buildLayeredImage doesn't dereference /nix/store symlink layers"` test always fails when I run it locally, so I comment it out when running the tests. But it seems to succeed in CI?

roberth commented:
Unfortunately we have to give ofborg a kick to make it test `dockerTools`. Let's see...

@ofborg build nixosTests.docker-tools

roberth commented:
Thanks!

The build succeeded on CI and on my machine, so it does indeed look like it could be something in the hardware or kernel or whatever. Unfortunate, but at least dockerTools seems to be ok.

WxNzEMof commented:
> Please cherry-pick the changes locally and resolve any conflicts.

OK, here it is: https://github.com/NixOS/nixpkgs/pull/302113
